Add an option to use `RenewTime` in leader election

AliDatadog opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR proposes an option to use the lease field `RenewTime` to choose if a leader should be replaced.
Currently, we use the `observedTime` for this decision which is tolerant to clock skew within a same cluster. It means that even if the lease has expired, candidates that observe it for the first time won't try to acquire it.
It could be observed with a simple new test case under `testTryAcquireOrRenew` which doesn't pass on the current master branch:
```
		{
			name: "first acquisition of expired object",
			reactors: []Reactor{
				{
					verb: "get",
					reaction: func(action fakeclient.Action) (handled bool, ret runtime.Object, err error) {
						return true, createLockObject(t, objectType, action.GetNamespace(), action.(fakeclient.GetAction).GetName(), &rl.LeaderElectionRecord{HolderIdentity: "bing", AcquireTime: metav1.NewTime(past), RenewTime: metav1.NewTime(past)}), nil
					},
				},
				{
					verb: "update",
					reaction: func(action fakeclient.Action) (handled bool, ret runtime.Object, err error) {
						return true, action.(fakeclient.CreateAction).GetObject(), nil
					},
				},
			},
			expectSuccess:    true,
			transitionLeader: true,
			outHolder:        "baz",
		},
```
This behavior is desirable in a scenario where the different nodes of a same cluster might be out of sync. However, for many use cases, we can assume that the clock skew is bounded.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

AliDatadog committed:
replace usage of observe time by acquire time in leader election

AliDatadog committed:
use renew instead and add case where the object is expired

alexzielenski commented:
I don't have context on this part of the code, but it seems like this is by design that the `RenewTime` on the record is not used:

https://github.com/kubernetes/kubernetes/blob/139fa6962026d1013d1d67e3f9f31d11da290199/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L22-L28

An earlier iteration of that in #77991 specifically mentions renew time on the record should not be used:
> A client observes timestamps captured locally to infer the state of the
> leader election. Thus the implementation is tolerant to arbitrary clock
> skew, but is not tolerant to arbitrary clock skew rate. Timestamp(renew time)
> is not meaningful if it was collected on another machine. The implementation
> of this client only acts on locally collected timestamps and cannot rely on
> the accuracy of timestamp in the record for correctness.

/cc @mikedanese 
/triage accepted

AliDatadog commented:
Thank you for the quick answer and for the clarification.

My change could break the current properties of the leader election, thus I believe we should perhaps hide it behind an optional flag.

We currently use this package for many applications and noticed that they can be slow to initialize because candidates don't try to acquire the expired leader token. Depending on the configuration, it can take a lot of time. Considering that the clock skew is often bounded within a cluster and that a recommended value for `LeaseDuration` is 60s, I believe an option to use `RenewTime` instead of `observedTime` could be relevant.

https://github.com/kubernetes/client-go/blob/6030d798e44096d5ffb58e00802cadd81a74dcf0/tools/leaderelection/leaderelection.go#L36

AliDatadog committed:
make the change optional

AliDatadog committed:
add ClockSkewToleration parameter

AliDatadog committed:
documentation update

AliDatadog commented:
@mikedanese I updated the PR to make this change optional and added the `ClockSkewToleration` parameter. In many scenarios, we can assume that the clock skew is bounded. 
We are using this package to elect a leader for one of our services and run a leader election every N seconds. However, on new redeployments, it is possible that the first election takes more time than expected because our new deployment does not try to acquire an expired lease.
